commit d3f8f24b6dc63f630a5a050496617aa0f4d29191
Author: Mrokkk <mrokk@protonmail.com>
Date:   Sun Apr 6 20:59:02 2025 +0200

    Some features
    
    * add logger
    * add support for 0xe9 debug console
    * add QT_NO_DEBUG_OUTPUT flag
    * fix for 5 seconds delay after closing app
    * workaround for double-free
    * do not use gamemode

diff --git a/.gitignore b/.gitignore
index 470a352ea..1cb6b5922 100644
--- a/.gitignore
+++ b/.gitignore
@@ -64,6 +64,7 @@ CMakeLists.txt.user
 
 # clangd
 .cache
+compile_commands.json
 
 # Jetbrains CLion
 .idea
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6338b635e..153f2fed9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -44,6 +44,16 @@ project(86Box
 include(CPack)
 include(CMakeDependentOption)
 
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+
+if(CMAKE_EXPORT_COMPILE_COMMANDS)
+    execute_process(
+        COMMAND ${CMAKE_COMMAND} -E create_symlink
+            ${CMAKE_BINARY_DIR}/compile_commands.json
+            ${CMAKE_SOURCE_DIR}/compile_commands.json
+    )
+endif()
+
 # Basic build options
 if(VCPKG_TOOLCHAIN)
     # For vcpkg builds we have to respect the linking method used by the
@@ -233,6 +243,8 @@ if(NOT EMU_COPYRIGHT_YEAR)
     set(EMU_COPYRIGHT_YEAR 2026)
 endif()
 
+add_compile_options(-ggdb3)
+
 # Libasan
 if(LIBASAN)
     add_compile_options(-fsanitize=address)
diff --git a/src/86box.c b/src/86box.c
index 4f5da88f1..788470cf1 100644
--- a/src/86box.c
+++ b/src/86box.c
@@ -1382,6 +1382,11 @@ pc_full_speed(void)
     atfullspeed = 1;
 }
 
+static void debug_outb(uint16_t, uint8_t val, void *)
+{
+    putc(val, stderr);
+}
+
 /* Initialize modules, ran once, after pc_init. */
 int
 pc_init_roms(void)
@@ -1727,6 +1732,8 @@ pc_reset_hard_init(void)
     /* Reset the general machine support modules. */
     io_init();
 
+    io_sethandler(0xe9, 1, NULL, NULL, NULL, &debug_outb, NULL, NULL, NULL);
+
     /* Turn on and (re)initialize timer processing. */
     timer_init();
 
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index c677c5367..c386086da 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -51,6 +51,8 @@ add_executable(86Box
     machine_status.c
 )
 
+add_compile_definitions("QT_NO_DEBUG_OUTPUT")
+
 if(CMAKE_SYSTEM_NAME MATCHES "Linux")
     add_compile_definitions(_FILE_OFFSET_BITS=64 _LARGEFILE_SOURCE=1 _LARGEFILE64_SOURCE=1)
 endif()
diff --git a/src/cpu/x86.c b/src/cpu/x86.c
index a262328b7..a35995834 100644
--- a/src/cpu/x86.c
+++ b/src/cpu/x86.c
@@ -35,6 +35,7 @@
 #include <86box/mem.h>
 #include <86box/rom.h>
 #include <86box/nmi.h>
+#include <86box/log.h>
 #include <86box/pic.h>
 #include <86box/pci.h>
 #include <86box/ppi.h>
@@ -237,6 +238,7 @@ makeznptable(void)
 static void
 reset_common(int hard)
 {
+    logger("CPU reset");
 #ifdef ENABLE_X86_LOG
     if (hard)
         x86_log("x86 reset\n");
diff --git a/src/cpu/x86seg.c b/src/cpu/x86seg.c
index 4e9b9d661..22ec79385 100644
--- a/src/cpu/x86seg.c
+++ b/src/cpu/x86seg.c
@@ -28,6 +28,7 @@
 #include "cpu.h"
 #include <86box/device.h>
 #include <86box/timer.h>
+#include <86box/log.h>
 #include <86box/machine.h>
 #include <86box/mem.h>
 #include <86box/nvr.h>
@@ -1548,6 +1549,14 @@ pmodeint(int num, int soft)
         x86gpf("pmodeint(): V86 banned int", 0);
         return;
     }
+
+    if (num < 14 && ((cpu_state.seg_cs.seg & 0x3) == 0)) {
+        logger("exception %u at %#x", num, cpu_state.pc);
+        if (num == 6) {
+            logger("  opcode: %#x", readmeml(0, cpu_state.pc));
+        }
+    }
+
     addr = (num << 3);
     if ((addr + 7) > idt.limit) {
         if (num == 0x08) {
diff --git a/src/include/86box/log.h b/src/include/86box/log.h
index c2c732495..65b78a97c 100644
--- a/src/include/86box/log.h
+++ b/src/include/86box/log.h
@@ -33,6 +33,7 @@ extern "C" {
 #define LOG_MINIMUM_REPEAT_ORDER        4               /* Minimum repeat size */
 
 /* Function prototypes. */
+extern void logger(const char* fmt, ...);
 extern void log_set_suppr_seen(void *priv, int suppr_seen);
 extern void log_set_dev_name(void *priv, char *dev_name);
 #ifndef RELEASE_BUILD
diff --git a/src/qt/qt_main.cpp b/src/qt/qt_main.cpp
index a99308ab1..7c1544c00 100644
--- a/src/qt/qt_main.cpp
+++ b/src/qt/qt_main.cpp
@@ -56,8 +56,8 @@ extern "C" {
 #include <86box/version.h>
 #include <86box/renderdefs.h>
 #ifdef Q_OS_LINUX
-#    define GAMEMODE_AUTO
-#    include "../unix/gamemode/gamemode_client.h"
+//#    define GAMEMODE_AUTO
+//#    include "../unix/gamemode/gamemode_client.h"
 #endif
 }
 
@@ -627,7 +627,7 @@ main(int argc, char *argv[])
     app.setStyle(new StyleOverride());
 
     bool startMaximized = window_remember && monitor_settings[0].mon_window_maximized;
-    fprintf(stderr, "Qt: version %s, platform \"%s\"\n", qVersion(), QApplication::platformName().toUtf8().data());
+    //fprintf(stderr, "Qt: version %s, platform \"%s\"\n", qVersion(), QApplication::platformName().toUtf8().data());
     Preferences::loadTranslators(&app);
 #ifdef Q_OS_WINDOWS
     QApplication::setFont(Preferences::getUIFont());
diff --git a/src/qt/xinput2_mouse.cpp b/src/qt/xinput2_mouse.cpp
index 5643df4d4..ccb827a9f 100644
--- a/src/qt/xinput2_mouse.cpp
+++ b/src/qt/xinput2_mouse.cpp
@@ -243,7 +243,7 @@ xinput2_exit()
 {
     if (!exitthread) {
         exitthread = true;
-        procThread->wait(5000);
+        procThread->wait(20);
         procThread->terminate();
     }
 }
diff --git a/src/utils/log.c b/src/utils/log.c
index 91f805d96..ee6daa0b4 100644
--- a/src/utils/log.c
+++ b/src/utils/log.c
@@ -51,6 +51,18 @@ extern FILE *stdlog;
 /* Functions only used in this translation unit. */
 void log_ensure_stdlog_open(void);
 
+void logger(const char* fmt, ...)
+{
+    va_list args;
+    char buffer[256];
+
+    va_start(args, fmt);
+    vsnprintf(buffer, sizeof(buffer), fmt, args);
+    va_end(args);
+
+    fprintf(stderr, "7,0,0.000000;[86Box] %s\n", buffer);
+}
+
 void
 log_set_dev_name(void *priv, char *dev_name)
 {
diff --git a/src/video/video.c b/src/video/video.c
index 600e59748..81dd38f72 100644
--- a/src/video/video.c
+++ b/src/video/video.c
@@ -842,7 +842,7 @@ video_monitor_close(int monitor_index)
         free(monitors[monitor_index].mon_pal_lookup);
     if (!monitors[monitor_index].mon_cga_palette_static)
         free(monitors[monitor_index].mon_cga_palette);
-    destroy_bitmap(monitors[monitor_index].target_buffer);
+    //destroy_bitmap(monitors[monitor_index].target_buffer); // FIXME: causes double free
     monitors[monitor_index].target_buffer = NULL;
     memset(&monitors[monitor_index], 0, sizeof(monitor_t));
 }
